test/unit/fiber_channel.cc

```diff
@@ -78,12 +78,167 @@ fiber_channel_get()
 	status = check_plan();
 }
 
+static void
+fiber_channel_close_basic(enum fiber_channel_close_mode mode)
+{
+	fiber_channel_set_close_mode(mode);
+
+	bool graceful = mode == FIBER_CHANNEL_CLOSE_GRACEFUL;
+	struct fiber_channel *channel = fiber_channel_new(10);
+
+	char msg_1;
+	char msg_2;
+	char msg_3;
+	void *ptr = NULL;
+
+	ok(fiber_channel_put_timeout(channel, &msg_1, 0) == 0,
+	   "fiber_channel_put(msg_1)");
+
+	ok(fiber_channel_put_timeout(channel, &msg_2, 0) == 0,
+	   "fiber_channel_put(msg_2)");
+
+	fiber_channel_get(channel, &ptr);
+
+	ok(ptr == &msg_1, "fiber_channel_get(1)");
+
+	fiber_channel_close(channel);
+
+	ok(channel->is_closed, "is_closed");
+	ok(channel->is_destroyed == !graceful,
+	   graceful ? "not is_destroyed" : "is_destroyed");
+
+	ok(fiber_channel_put_timeout(channel, &msg_3, 0) != 0,
+	   "not fiber_channel_put(msg_3)");
+
+	ptr = NULL;
+	fiber_channel_get(channel, &ptr);
+
+	ok(ptr == (graceful ? &msg_2 : NULL),
+	   graceful ? "fiber_channel_get(1)" : "not fiber_channel_get(1)");
+	ok(channel->is_destroyed, "is_destroyed");
+
+	fiber_channel_delete(channel);
+}
+
+static int
+reader_f(va_list ap)
+{
+	struct fiber_channel *channel = (struct fiber_channel *)fiber()->f_arg;
+	void *msg;
+	ok(!channel->is_closed, "reader_f tries to read from the open channel");
+	/*
+	 * Try to obtain the message into the zero-length channel.
+	 * XXX: <reader> fiber hangs forever, until one of the
+	 * following occurs:
+	 * * <fiber_channel_put> is called from another fiber
+	 * * <channel> is closed from another fiber
+	 * For the latter case <fiber_channel_get> fails (i.e.
+	 * yields non zero status).
+	 */
+	ok(fiber_channel_get(channel, &msg) != 0,
+	   "reader_f fails to read a message from the zero-length channel");
+	ok(channel->is_closed, "reader_f hangs until channel is closed");
+	ok(fiber_channel_get_timeout(channel, &msg, 0) != 0,
+	   "reader_f fails to read a message from the closed channel");
+	ok(msg == NULL, "reader_f obtains NULL from the closed channel");
+
+	return 0;
+}
+
+static void
+fiber_channel_close_reader(enum fiber_channel_close_mode mode)
+{
+	fiber_channel_set_close_mode(mode);
+
+	bool graceful = mode == FIBER_CHANNEL_CLOSE_GRACEFUL;
+	struct fiber_channel *channel = fiber_channel_new(1);
+	struct fiber *reader = fiber_new("reader", reader_f);
+	fail_if(reader == NULL);
+	reader->f_arg = channel;
+	fiber_wakeup(reader);
+	/* Yield to start tests in <reader> payload. */
+	fiber_sleep(0);
+
+	fiber_channel_close(channel);
+	ok(channel->is_closed, "is_closed");
+	/* Yield to finish tests in <reader> payload. */
+	fiber_sleep(0);
```

Comment:
Can we at least assert that the `reader_f()` function reaches its end? We can either set a global variable or make the fiber joinable and join it.

Now we lean on particular fiber scheduling algorithm and it feels fragile.

Reply:
Sounds reasonable. Replaced this `fiber_sleep` with `fiber_join`.